### PR TITLE
Fix pathlib compatibility in file I/O

### PR DESCRIPTION
**Pull request by duncanmmacleod**
This PR fixes up a number of functions to enable reading from `pathlib.Path` objects as well as `file` and `str`. This probably supersedes #1360.

### CODE REVIEW

**Comment by duncanmmacleod** [comment]
Ping @gwpy/maintainers, please review this.

File: gwpy/io/utils.py [hunk 1]
```diff
@@ -98,6 +100,50 @@ def gopen(name, *args, **kwargs):
...
+
+    >>> @with_open
+    >>> def my_func(pathorfile, *args, **kwargs)
+    >>>     ...
+
+    To ensure that the second argument (position 1) is a file open for writing:
+
+    >>> @with_open(mode="r", pos=1)
+    >>> def my_func(stuff, pathorfile, *args, **kwargs)
+    >>>     stuff.write_to(pathorfile, *args, **kwargs)
```
Comment (eagoetz) [inline gwpy/io/utils.py:130]: This example is a bit confusing: which is readable (`pathorfile`?) and which is writable(`stuff`?)? Perhaps give a bit more explanation here for non-experts.

File: gwpy/io/utils.py [hunk 2]
```diff
@@ -98,6 +100,50 @@ def gopen(name, *args, **kwargs):
...
+            if not isinstance(args[pos], FILE_LIKE):
+                with open(args[pos], mode=mode) as fobj:
+                    args = list(args)
+                    args[pos] = fobj
+                    return func(*args, **kwargs)
+            return func(*args, **kwargs)
+        return wrapped_func
+    if func:
+        return _decorator(func)
+    return _decorator
```
Comment (eagoetz) [inline gwpy/io/utils.py:144]: Can you add some in-line code comments here? We're not able to follow exactly what is going on.

File: gwpy/io/utils.py (hunk 1 above)
Comment (eagoetz) [reply gwpy/io/utils.py:130]: It looks like mode in the example should be `w`, or `rw`, not `r` (unless we don't understand something fundamental...)

**Comment by eagoetz** [comment]
Trying to leave comments but github is preventing us from leaving comments. We will come back to this

**Comment by eagoetz** [comment]
@duncanmmacleod We think this is basically fine, though our understanding of exactly the changes to `gwpy/io/tests/test_utils.py` are not well understood, but there is a potential bug in `gwpy/detector/io/clf.py` line 141 (see comment).

File: gwpy/io/utils.py (hunk 2 above)
Comment (duncanmmacleod) [reply gwpy/io/utils.py]: I've added inline comments to explain the logic internal to the decorated function, the outer scaffolding is just the standard structure to define a decorator that can be called with or without functional brackets, i.e. allowing both

```python
@with_open
def thing():
    ...
```

and

```python
@with_open(mode="w")
def thing():
    ...
```

**Comment by duncanmmacleod** [comment]
> @duncanmmacleod We think this is basically fine, though our understanding of exactly the changes to `gwpy/io/tests/test_utils.py` are not well understood,

That change is just pytest fanciness, there isn't actually anything substantive in the changes to what the tests actually do.

> but there is a potential bug in `gwpy/detector/io/clf.py` line 141 (see comment).

This has been corrected, good catch.

I hope this is now ready to go.

**Review by eagoetz — APPROVED** [review]
LGTU

File: gwpy/detector/io/omega.py [hunk 3]
```diff
@@ -56,30 +57,18 @@ def read_omega_scan_config(source):
...
+        if not line or line.startswith('#'):
+            continue
+        if line.startswith('['):
+            section = line[1:-1]
+        elif line.startswith('{'):
+            append(parse_omega_channel(source, section))
+        else:
+            raise RuntimeError(
+                "Failed to parse Omega config line: {!r}".format(line),
+            )
```
[the remark this answers is not in the document]
Comment (eagoetz) [reply gwpy/detector/io/omega.py:71]: @duncanmmacleod just want to make sure you saw this

**Pull request merged**